### PR TITLE
Changes input record sn nonce allocation, adds noop empty payload check

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR requires resampling parameters.

### Inner circuit changes

- [-2 constraints] Changes the input record's serial number nonce allocation:
```rs
// Before (pseudocode)

let serial_number_nonce = Circuit::alloc(record.serial_number_nonce()); // never used again.
let serial_number_nonce_bytes = Circuit::alloc(serial_number_nonce.to_bytes_le());

// After (pseudocode)

let serial_number_nonce_bytes = Circuit::alloc(record.serial_number_nonce().to_bytes_le());
```
- [+4096 constraints] Adds noop record empty payload checks (in addition to existing zero value checks):
```rs
given_payload.conditional_enforce_equal(cs, empty_payload, given_is_dummy);
```
- [-6 constraints] Changes the zero value alloc to constant.

### General changes

- Adds `Transaction::to_encrypted_record_hashes()`
- Adds `TransactionKernel::to_output_serial_number_nonces()`
- Adds `InnerPublicVariables::new()`
- Adds `Parameters::SerialNumberNonce` convenience type

### CODE REVIEW

File: dpc/src/circuits/inner_circuit.rs [hunk 1]
```diff
@@ -148,7 +147,11 @@ pub fn execute_inner_circuit<C: Parameters, CS: ConstraintSystem<C::InnerScalarF
         )
     };
 
-    let zero_value = UInt8::alloc_vec(&mut cs.ns(|| "Declare record zero value"), &to_bytes_le![0u64]?)?;
+    let zero_value = UInt8::alloc_vec(&mut cs.ns(|| "Declare a record zero value"), &(0u64).to_bytes_le()?)?;
```
Comment (weikengchen) [inline dpc/src/circuits/inner_circuit.rs:150]: Shouldn't it be `constant_vec`?

File: dpc/src/circuits/inner_circuit.rs [hunk 2]
```diff
@@ -148,7 +147,11 @@ pub fn execute_inner_circuit<C: Parameters, CS: ConstraintSystem<C::InnerScalarF
...
+    let empty_payload = UInt8::alloc_vec(
```
Comment (weikengchen) [inline dpc/src/circuits/inner_circuit.rs:151]: Shouldn't it be `constant_vec`?

**Review by weikengchen — APPROVED** [review]

File: dpc/src/circuits/inner_circuit.rs (hunk 1 above)
Comment (howardwu) [reply dpc/src/circuits/inner_circuit.rs:150]: Good catch

File: dpc/src/circuits/inner_circuit.rs (hunk 2 above)
Comment (howardwu) [reply dpc/src/circuits/inner_circuit.rs:151]: Good catch

File: dpc/src/circuits/inner_circuit.rs (hunk 1 above)
Comment (howardwu) [reply dpc/src/circuits/inner_circuit.rs]: Fixed

File: dpc/src/circuits/inner_circuit.rs (hunk 2 above)
Comment (howardwu) [reply dpc/src/circuits/inner_circuit.rs]: Fixed

**Comment by howardwu** [comment]
Resampled inner circuit parameters

**Comment by Pratyush** [comment]
Sorry I didn't comment on this earlier, but to reduce the cost of the payload = 0, check maybe we should pack the bytes into field elements, and check that those field elements are zero?

**Comment by weikengchen** [comment]
Agree, it will be cheaper. 

It is unlikely that we change the underlying bit comparison implementation (there is a generic implementation for Vec<_>), but we can do it manually here, by packing them to field elements first.

**Pull request merged**